tests/test_markers.py

```diff
@@ -388,12 +387,38 @@ def test_extra_str_normalization(self):
         assert str(Marker(rhs)) == f'extra == "{normalized_name}"'
 
     def test_python_full_version_untagged_user_provided(self):
-        """A user-provided python_full_version ending with a + fails to parse."""
-        with pytest.raises(InvalidVersion):
-            Marker("python_full_version < '3.12'").evaluate(
-                {"python_full_version": "3.11.1+"}
-            )
+        """A user-provided python_full_version ending with a + uses Python behaviour."""
+        env = {"python_full_version": "3.11.1+"}
+        assert Marker("python_full_version < '3.12'").evaluate(env)
 
     def test_python_full_version_untagged(self):
         with mock.patch("platform.python_version", return_value="3.11.1+"):
             assert Marker("python_full_version < '3.12'").evaluate()
+
+    @pytest.mark.parametrize(
+        ("marker_string", "environment", "expected"),
+        [
+            ("platform_release >= '20.0'", {"platform_release": "21-foobar"}, True),
+            ("platform_release >= '8'", {"platform_release": "6.7.0-gentoo"}, False),
+            ("platform_version == '27'", {"platform_version": "weird string"}, False),
+            (
+                "implementation_version == '3.*'",
+                {"implementation_version": "2_private"},
+                False,
+            ),
+            (
+                "implementation_version == '3.*'",
+                {"implementation_version": "3.*"},
+                True,
+            ),
+        ],
+    )
+    def test_valid_specifier_invalid_version_fallback_to_python(
+        self, marker_string: str, environment: dict, expected: bool
+    ):
+        """If the right operand is a valid version specifier, but the
+        left operand is not a valid version, fallback to Python string
+        comparison behaviour.
+        """
+        args = [] if environment is None else [environment]
```

Comment:
`environment` is typed as a `dict` here so you shouldn't have to account for it being `None`, right?

Reply:
I kept it in as I thought someone may want to add a test case that doesn't specify an environment override later, but that doesn't really make sense here, so I've removed this line. Good catch!

Reply:
Makes sense! Changing the declared type of the argument to `Optional[dict]` (or maybe `Union[dict, None]` is the preferred spelling? I forget :shrug:) would also be totally reasonable as far as I'm concerned, if you wanted to go that direction.